### PR TITLE
ADD: add passphrase support to wallets

### PR DESCRIPTION
**Pull request by limpbrains**
Questions: 
- do I need to refactor `HDAezeedWallet` to store password as other hd wallets, in this.passphrase? I don't like the idea to store passphrase in the same string, divided by colon.
- could wallets with passphrase be used in Multisig?

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/mv60juggct26rkh6405fj1a07r?device=pixel4

[download apk](http://filestorage.bluewallet.io:1488/limpbrains-passphrase-1621010345.apk)

**Review by jaonoctus — APPROVED** [review]
tACK

```
HDLegacyElectrumSeedP2PKHWallet
    ✓ can use mnemonic with passphrase (4 ms)
Legacy HD (BIP44)
    ✓ can use mnemonic with passphrase (4 ms)
Bech32 Segwit HD (BIP84)
    ✓ can use mnemonic with passphrase (6 ms)
HDSegwitElectrumSeedP2WPKHWallet
    ✓ can use mnemonic with passphrase (6 ms)
P2SH Segwit HD (BIP49)
    ✓ can use mnemonic with passphrase (4 ms)
SLIP39 wallets tests
    ✓ SLIP39LegacyP2PKHWallet can use passphrase (15 ms)
    ✓ SLIP39SegwitP2SHWallet can use passphrase (7 ms)
    ✓ SLIP39SegwitBech32Wallet can use passphrase (8 ms)
```

**Comment by jaonoctus** [comment]
@limpbrains  will you include the `setPassphrase` function on the [import screen](https://github.com/BlueWallet/BlueWallet/blob/44522b45af4f1850cb7e85d0d634075ae905d8cd/screen/wallets/import.js) later?

**Comment by limpbrains** [comment]
> @limpbrains will you include the `setPassphrase` function on the [import screen](https://github.com/BlueWallet/BlueWallet/blob/44522b45af4f1850cb7e85d0d634075ae905d8cd/screen/wallets/import.js) later?

Yes I will in another PR. This one is just a draft to discuss wallets and API.
@Overtorment plz take a look

**Comment by Overtorment** [comment]
sorry for the delay!
okay, Im conceptualy onboard with the change, and storing passphrase in a separate field. lets refactor aezeed to work the same way.

as for multisig, I've never seen an option to have passwords for cosigners. might be wrong tho.


PS. remind me, passphrase for BIP39 mnemonic seeds, they are not purely passphrases? there is no way to tell if given BIP39 seed needs a passphrase or not? with an incorrect passphrase it will just unwrap into empty wallet, right?

**Review by Overtorment — CHANGES_REQUESTED** [review]
conceptually lgtm

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/x1d93bmkqdpar6h0yeqxc0nzr8?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/7246)

**Comment by limpbrains** [comment]
> PS. remind me, passphrase for BIP39 mnemonic seeds, they are not purely passphrases? there is no way to tell if given BIP39 seed needs a passphrase or not? with an incorrect passphrase it will just unwrap into empty wallet, right?

Correct

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/0m1m0gfxqq5qg81wt7hdvrd1hw?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/7247)

**Comment by Overtorment** [comment]
is it ready for review & merge?

**Comment by limpbrains** [comment]
> is it ready for review & merge?

Not yet. I need to figure out where to put code to migrate old aezeed wallets to new data structure. And maybe change some apis.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/r7b20em2kdrmgkfc7ndbcd9vq4?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/7281)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/p5erezy4kmu12h56n67mgnyhrg?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/7282)

**Comment by limpbrains** [comment]
ready to be merged, I will start working on improving import screen in another PR

File: class/app-storage.js [hunk 1]
```diff
@@ -364,6 +364,14 @@ export class AppStorage {
             break;
           case HDAezeedWallet.type:
             unserializedWallet = HDAezeedWallet.fromJson(key);
+            // migrate password to this.passphrase field
+            // remove this code somewhere in year 2022
```
Comment (Overtorment) [inline class/app-storage.js:368]: there is nothing more permanent then temporary solution kek

Comment (limpbrains) [reply class/app-storage.js:368]: True :) I should probably add // TODO and FIXME

File: tests/unit/hd-aezeed.test.js [hunk 2]
```diff
@@ -11,40 +11,47 @@ describe('HDAezeedWallet', () => {
...
     );
+    aezeed.setPassphrase();
     assert.ok(await aezeed.validateMnemonicAsync());
     assert.ok(!(await aezeed.mnemonicInvalidPassword()));
 
     aezeed.setSecret(
-      'able concert slush lend olive cost wagon dawn board robot park snap dignity churn fiction quote shrimp hammer wing jump immune skill sunset west:aezeed',
+      'able concert slush lend olive cost wagon dawn board robot park snap dignity churn fiction quote shrimp hammer wing jump immune skill sunset west',
     );
+    aezeed.setPassphrase();
```
Comment (Overtorment) [inline tests/unit/hd-aezeed.test.js:46]: extra line

**Review by Overtorment — APPROVED** [review]
image
image

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!

**Comment by Overtorment** [comment]
very noice.
we could include it in next release.
as a workaround while there is no gui, we could accept text with semicolon on import screen (`seed:pass`)
unless we do that already..?

**Comment by limpbrains** [comment]
> very noice.
> we could include it in next release.
> as a workaround while there is no gui, we could accept text with semicolon on import screen (`seed:pass`)
> unless we do that already..?

I'm going to deliver UI pretty soon

File: class/wallets/abstract-hd-wallet.js [hunk 3]
```diff
@@ -61,17 +62,21 @@ export class AbstractHDWallet extends LegacyWallet {
...
+  }
+
   /**
    * @return {Boolean} is mnemonic in `this.secret` valid
    */
   validateMnemonic() {
     return bip39.validateMnemonic(this.secret);
   }
 
-  getMnemonicToSeedHex() {
```
[the remark this answers is not in the document]
Comment (limpbrains) [reply class/wallets/abstract-hd-wallet.js:71]: Looks like so, I can't remember why I removed it 😕

**Pull request merged**